lib/matplotlib/figure.py

```diff
@@ -2084,6 +2088,166 @@ def tight_layout(self, renderer=None, pad=1.08, h_pad=None, w_pad=None,
             pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
         self.subplots_adjust(**kwargs)
 
+    def align_xlabels(self, axs=None):
+        """
+        Align the ylabels of subplots in the same subplot column if label
+        alignment is being done automatically (i.e. the label position is
+        not manually set).
+
+        Alignment persists for draw events after this is called.
+
+        If a label is on the bottom, it is aligned with labels on axes that
+        also have their label on the bottom and that have the same
+        bottom-most subplot row.  If the label is on the top,
+        it is aligned with labels on axes with the same top-most row.
+
+        Parameters
+        ----------
+        axs : list of `~matplotlib.axes.Axes` (None)
+            Optional list of (or ndarray) `~matplotlib.axes.Axes` to align
+            the xlabels.  Default is to align all axes on the figure.
+
+        Note
+        ----
+        This assumes that ``axs`` are from the same `~.GridSpec`, so that
+        their `~.SubplotSpec` positions correspond to figure positions.
+
+        See Also
+        --------
+        matplotlib.figure.Figure.align_ylabels
+
+        matplotlib.figure.Figure.align_labels
+
+        Example
+        -------
+        Example with rotated xtick labels::
+
+            fig, axs = plt.subplots(1, 2)
+            for tick in axs[0].get_xticklabels():
+                tick.set_rotation(55)
+            axs[0].set_xlabel('XLabel 0')
+            axs[1].set_xlabel('XLabel 1')
+            fig.align_xlabels()
+
+        """
+
+        if axs is None:
+            axs = self.axes
+        axs = np.asarray(np.array(axs)).flatten().tolist()
```

Comment:
You really really wanted to make sure that you got a list heh :-)
Currently this will also accept Axes instances (rather than Axes in a single-element list) and multi-dimensional arrays.  Was that intended?  Just making sure, I would as usual prefer a simpler interface but am only -0 on being more flexible.

Reply:
Too much?

Ummm, I can't remember why I went so nuts on this.  

  - I agree a singleton would be silly, though letting a list w/ one element would be fine.  
  - It needs to be flattened, but before I can do that, it needs to be an np.array... 
  - but, I have no idea why I have `np.array` and `np.asarray` in there.

Fixed?  
```python
axs = np.asarray(axs).flatten().tolist()
```

Reply:
`for ax in np.asarray(axs).flat:` is enough (that'll give you an 1D iterator that directly refers to the original data).

Reply:
Couldn't quite get that to work, and I want to save the list so I don't need the awkward construct  twice.  Using `axs = np.asarray(axs).flatten()`.

Reply:
TO be honest, not sure why `flat` doesn't work.  It basically dropped one out of three axes on the figure in `align_labels_demo.py` so that third axis didn't get aligned.  I don't really get why it doesn't work, but `flatten()` does, so...

Reply:
Ah, I missed the fact that you had a nested loop also iterating over axs.  In that case it is normal that the inner loop also advances the outer iterator and that won't work indeed.

As a side side note, using `.ravel()` instead of `.flatten()` is usually more efficient, because the latter always copies the data into a new array, whereas the former only does so when needed (i.e. when the data of the original array is not regularly spaced in memory).  We're unlikely to ever have so many axes to run out of memory :-) but still a good habit IMO.

Reply:
Ah, I see.   Interesting.  I never understood the difference between `flatten()` and `ravel()` and just use `flatten` because the word means something to me, whereas I'm not sure what `ravel` is supposed to mean.

Reply:
"ravel" is apparently a synonym of "unravel" :) https://www.merriam-webster.com/dictionary/ravel?utm_campaign=sd&utm_medium=serp&utm_source=jsonld
[OT: reminds me of the day where my colleague explained to me that "being down for sthg" and "being up for sthg" meant essentially the same thing.  English is sometimes weird :)]

Reply:
Made even better by numpy's use of `unravel` to mean the opposite of `ravel`: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.unravel_index.html